### ISSUE TITLE
Recovering lost tabs after extension update or removal

### ISSUE THREAD

**Issue by deanoemcke** [deanoemcke/thegreatsuspender#526]
This is the official issue for users trying to recover from tab loss due to an update of the extension, or following removal/disabling of the extension. This is the number one pain point of the extension and here I will attempt to address why this is an issue, and what you can do if affected by it.

**Why do my tabs disappear when the extension updates or is removed?**

The Great Suspender works by redirecting a tab to a new url in order to 'suspend' it. This means that the tab is now controlled by the extension process. When the extension updates or is disabled or uninstalled, this process is killed and all tabs that belong to it are removed from the browser. I have an on-going policy to *never* force automatic updates of this extension when publishing a new version on the chrome webstore. However, sometimes this is out of my control. 

For more information on my 'no updates' policy, please refer to this thread: https://github.com/deanoemcke/thegreatsuspender/issues/322

It should be noted here that the extension *does* come with an inbuilt tab recovery system that will automatically detect and reload lost tabs should ever force the extension to be automatically updated. I have received feedback that this system does work, but there are edge cases where it fails.

**What is the safe way to remove the extension?**

If you want to uninstall the extension, please *unsuspend all tabs* before doing so. This is the only way to prevent those tabs from disappearing. This can be done easily by clicking the 'Unsuspend all tabs' option in the extension popup menu. Or more manually by visiting every suspended tab and manually reloading it.
Please note that if using the 'Unsuspend all tabs' option, you will need to do this once for *each chrome window* you have open.

If you failed to unsuspend all tabs before uninstalling and have lost tabs, please refer to the section below entitled "How to recover lost tabs without The Great Suspender".

Please note, uninstalling the extension will also permanently remove all extension data including tab history and extension options. Reinstalling the extension will not enable you to do any sort of recovery.

I would recommend anyone wanting to remove the extension to first back up their tabs using another extension called "[Session buddy](https://chrome.google.com/webstore/detail/session-buddy/edacconmaakjimmfgnblocblbcdcpbko?hl=en)". This tool will allow you to back up all your tabs and restore them again at a later date. Please be aware that those tabs suspended at the time the session buddy backup is performed will not have their correct urls. These links will only work as long as The Great Suspender is currently installed on your browser. If you want the real urls in your session buddy backup, then you will need to unsuspend all your tabs first.

**What is the safe way to update the extension?**

Unfortunately chrome does not give the user the ability to manage their own extension updates. As soon as a new release is made available on the webstore, this update is automatically pushed to users at an uncontrolled time. However, I do have the ability to disable this automatic updating. I *can* publish a new version on the webstore without it being pushed out to all users. But this means that existing users will never receive updates to the extension.

To safely update the extension to the latest webstore version, you must uninstall and then reinstall the extension. Beware that doing so will cause you to lose all suspended tabs, tab session history, extension settings, and whitelist information. So we must make sure to back these up first.

First, copy and paste the contents of your whitelist into a text file so you can paste it back in after the update. 

Secondly, we need to back up the tabs. See the section above entitled "What is the safe way to remove the extension?" for instructions on how to backup your tabs and remove the extension safely. If you are using session buddy to backup your tabs, and you are intending to reinstall The Great Suspender after the uninstall, then it should be safe to include suspended tabs in your session buddy backup. When restoring these tabs, they will be be restored in their suspended state (useful if you are restoring a session with 100+ tabs).

**What should I do if I have lost tabs?**

If you have lost tabs due to the extension being removed then refer to the section below entitled "How to recover lost tabs *without* The Great Suspender".

If you have lost tabs due to the extension being disabled, then first re-enable the extension, and then refer to the section below entitled "How to recover lost tabs with The Great Suspender".

If you have lost tabs but the extension still seems to be installed and running, then refer to the section below entitled "How to recover lost tabs with The Great Suspender".

**How to recover lost tabs with The Great Suspender**

The extension comes with its own tab history management UI to help users recover from lost tabs. Go to the extension options page (from 'settings' in the popup or 'options' when right-clicking on the extension). Then in the settings sidebar click on 'Session management'. This will show you your most recent tab sessions. You can click on each session to see more detail on the individual windows and tabs it contains.

To reload a session, simply click the 'reload' link. This will reload all windows and tabs in an 'unsuspended' state. If your session contains a very large number of tabs, then you might instead want to click 'resuspend' which will be much faster as it reloads the tabs in a suspended state.

If for some reason the missing tabs are not in your recent sessions, then please follow the guide below for recovering lost tabs without using The Great Suspender.

**How to recover lost tabs without The Great Suspender**

You can attempt to recover lost tabs by using chromes in-built history page. Navigate to chrome://history in a new tab and you will be shown a list of tabs you have visited in the past grouped by date and showing the most recent at the top. Somewhere in this list you will have a record of all the tabs you lost. However it can be a bit tricky to find them as they are mixed in with all the tabs you have visited and purposely closed as well.

For example, if you opened a tab one week ago, and it got suspended and you never revisited that tab, then in chrome history, it will be grouped with all the tabs from one week ago.

If you do find a lost tab in this list, there is a chance that when you try to reopen it, it will take you to a blank page saying "This site cannot be reached". Please refer to the section below on how to recover these tabs.

**My suspended tab says "This site cannot be reached"**

It can happen that when you open a suspended tab link, or try to unsuspend a tab, you will see a blank page with the text "This site cannot be reached". And it will have a strange url that looks something like this: `chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html#ttl=Google&uri=https://www.google.com`.

This is most likely due to the fact that you no longer have The Great Suspender installed in your browser. The easiest way to recover these tabs is to reinstall the extension, and then reload the page.

Should this fail for any reason (which would happen if you tried to open this url in another browser like firefox, or from a device that does not support extensions such as an android phone), then as a last resort you can manually edit the url to recover the tab. Delete everything before the &uri= text in the address bar and the page should reload corrently.
ie: in the example above you would end up with `https://www.google.com`.

**Comment by adcurtin** [comment]
will clicking "update extensions now" in the chrome extensions page update to the latest webstore version even if auto updates are disabled?

**Comment by deanoemcke** [comment]
@adcurtin I don't believe so. i prevent automatic updates by setting the 'max deploy percentage' to zero.
As detailed here: https://developer.chrome.com/webstore/publish
It is slightly ambiguous, but it does state: "The max deploy percentage control only applies to auto-update for existing users. New users always get the latest version of your app."
I think this also means that existing users are literally unable to receive updates. This is why I suggest uninstalling and reinstalling as the only way around this.

**Comment by uecasm** [comment]
What about https://developer.chrome.com/extensions/runtime#event-onUpdateAvailable?  It sounds like that would let you detect an impending update.

**Comment by deanoemcke** [comment]
@uecasm I can't believe I have never noticed this event. It's probably exactly what I need. I'll try it out. Thanks for the heads up! Unfortunately, it's going to take another update to get it implemented :(

**Comment by mengsel** [comment]
Is there any way to recover 'session management' sessions without the extension? I'd love to be able to export them to my new device...

**Comment by deanoemcke** [comment]
@mengsel In the session management page, there is a link to export each session. This produces a text file with each (unsuspended) url on a new line.
The Great Suspender has no way to import these sessions on another computer, but you can do exactly that with Session Buddy, another chrome extension which I mention in the information above.

**Comment by rudolphos** [comment]
Is there a way to delete this line from a list of URLS ?

`chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html#ttl= ... &uri=` DIRECT URL HERE

I can't do it anymore, because devs of this plugin have added #ttl= which has different text in every tab, previously I just removed the extension string and I got direct links.

**Comment by deanoemcke** [comment]
@rudolphos what exactly are you trying to do? if you want to export a list of urls in all your tabs, you can do exactly that from the session management screen within the extension (as mentioned in my reply to @mengsel above)

**Comment by rudolphos** [comment]
@deanoemcke When tab is suspended it now has a longer URL including tab title, which messes up 'order by url' feature in SessionBuddy if there are also other unsuspended tabs, which can cause duplicates if sessions are merged. This is important for me as I have sessions with 180+ tabs and many more saved and modified, merged.

I usually export Session URL list as CSV from sessionbuddy and previously it was easier to remove the sessionbuddy uri= part, but now that there are various titles in it, I had to make regex which seems to be working on one-per-line tab list, but it didn't work on db table exported from SessionBuddy.

**Comment by zxweed** [comment]
The extension are lost all my tabs and cannot recover from history, I am so sad, it was collect a couple of months...

**Comment by owenblacker** [comment]
The "Session management" tab is empty; the extension does not appear to be saving my sessions. Is there anything I can do to rectify this — automatically or manually?

**Comment by deanoemcke** [comment]
@owenblacker if you open the chrome developer console on the session management screen, are there any errors on the page? If so, could you create a new issue for this?

**Comment by owenblacker** [comment]
@deanoemcke I only just saw this notification and, while it's been continuously the case for some time now, I just went to the Session management tab and it is saving my sessions again now. I'm afraid I have no idea what's changed between those two :(

**Comment by epixcz** [comment]
@zxweed I can recommend Session Buddy. Its awesome, only downside is no synchronization between installations, but that can be done manually by export/import.

**Comment by hugegit** [comment]
Acckkk - I just lost a saved tab session by installing the current Alpha version (because of the Error 400 / gsScrollPos / Google issue) ... I thought saving it was enough and didn't know I had to export it. Is there any way to get it back? 

[whimper]

**Comment by fwextensions** [comment]
@hugegit loading the unpacked alpha version shouldn't affect the webstore version.  I've currently got both running (with only one set to auto-suspend tabs, to reduce confusion).  If you uninstalled the webstore version and then installed the alpha, I'm not sure if recovery is possible.  Maybe there's a backup of the indexeddb file that @adcurtin mentions above?

**Comment by danilaml** [comment]
So the GS won't update automatically when 7.0 comes, so that you can unsuspend all tabs before forcing an update? Did I understand correctly? Also, will the old "suspended" links work? I.e. if I have them stored in some session management extension, will they open and work correctly after the update?

**Comment by adcurtin** [comment]
I ran into an interesting problem related to updating.

I have multiple computers connected to my chrome account. I upgraded my main computer to 7.0, and disabled the webstore version. Well, disabling the webstore version synced to my other computers, and disabled the extension there as well. ~110 extension tabs were closed.

**Comment by deanoemcke** [comment]
@adcurtin yes. that is expected. re-enabling it should restore all your tabs automatically?
if you want that setup, then you're best to just set the v6.30 on your main computer to 'time to suspend -> never' which will essentially disable it just on that computer.

**Comment by BrentNewland** [comment]
Session Buddy sucks. It's never done a good job auto saving sessions for me. It either doesn't save them until I open it, in which case it saves immediately, or it doesn't save them on a regular basis.

I recommend Tabs Backup and Restore. It's something like 6 years old but still works perfectly. It can be set to backup anywhere from once a minute to once every four hours, and can save up to 80 backups. And IT ALWAYS WORKS.

**Comment by garrett-eclipse** [comment]
Hi @deanoemcke you mentioned that on re-enabling the extension your tabs should be restored. I think I found a case this isn't true. I'm logged into chrome from my laptop and desktop. I disabled the extension via my laptop and then arriving at work today enabled via the desktop. Now all suspended tabs on both laptop and desktop are gone. disabling and re-enabling from laptop doesn't help and the sessions don't hold the missing tabs. Thankfully i have Recently Closed Tabs extension installed so I can recover but just wanted to flag the case where you can disable from one machine and re-enable from another and in that case it won't automatically restore tabs.
Other than this crux of an issue I'm very grateful for your plugin.
Cheers

**Comment by dragonspirit1185** [comment]
@deanoemcke You should have never released the update if this wasn't fixed. I'm losing too many tabs. If you can not fix this then why not find someone who can? I liked it better when I just had to clear cookies. at least then I didn't lose tabs. I never even knew there was a tab limit. Apparently there is.
You obviously need to outsource and get this fixed.  
I liked it better when I just had to clear cookies. I wasn't losing tabs then. This update fixes one issue only to create and even bigger issue. 

I just used the restore session option and wow made my junk freeze for a while. I just wanted to restore one window. Make it to where we can choose each Tab or each window like you can with the extension One Tab. 
I don't need all this opening all at once.

**Comment by adcurtin** [comment]
@dragonspirit1185 it sucks that you lost tabs, and I'm sorry. I know how frustrating it can be. However, your last comment is not appropriate. Dean has provided this awesome extension for free, and has been very careful to avoid this problem as much as possible. Chrome does not make it easy.

Please do not come to the github and complain and whine like that, as it is not productive. Being an ass to an individual providing a service to you and receiving nothing in return is very poor form.

**Comment by dragonspirit1185** [comment]
It's feedback. I'm not whining. Free or not you gotta roll with the
criticisms.
I'm not being an "ass", I'm sharing my frustrations.

**Comment by dragonspirit1185** [comment]
You want to talk about being unproductive yet you're being nothing but
unproductive here so please move along.
You've stated your opinion enough.

Calling someone an "ass" and acting the way that you are towards them it's
the same thing well it's worse.
Two wrongs don't make a right.
You're confronting me in a much worse way than I expressed myself.
You have been profane and blatantly rude.
I actually want trying to be rude. I just happen to have a developmental
disability that causes me not to set how I come across when I'm frustrated.
I actually have autism that causes me to like appropriate social convention
and not really know how to word things.
If I was in his shoes I would definitely outsource.
I will not respond or entertain this any further.

**Comment by liamjohnston** [comment]
@dragonspirit1185 Open source software is essentially outsourced by definition. Either fix the problem yourself (as a pull request) or find someone else who can. 

@adcurtin Is right; your comments were unhelpful an inappropriate. The developer is doing everything he can to avoid this not-completely-avoidable problem.

**Comment by deanoemcke** [comment]
@dragonspirit1185 i'm not sure what you are referring to when you say "if this wasn't fixed?". you mean potential tab loss when updating? is that the "even bigger issue" you refer to? just to be clear, this is not an issue that has been created in the most recent release. it has been an ongoing issue every time i push an update to the chrome webstore. i'm also not sure what you mean by the "tab limit". can you clarify?

due to the way the extension works, there is absolutely nothing i can do about suspended tabs disappearing during an update. the code that restores tabs after the update seems to work well for almost everybody, but there seems to be always a small percentage of people for whom it fails.

in regards to your usage of the session manager. perhaps you haven't realised, but you can actually just reopen a single window from a session instead of the whole session. you need to expand the session containing the window you want to restore, when next to the 'window 1' (or whatever number), click the 'resuspend' button. you could alternatively click the 'reload' button, but this will cause every tab in the window to load properly. from the sounds of it, that's what you've been doing. because if you choose 'resuspend' instead, it can actually open a window with quite a large number of tabs pretty quickly.

i'm sorry that i've caused a loss of tabs for you. i do understand how frustrating that can be. as you know, the impetus for this release was to fix the youtube cookies bug. and i made a call that it was more important to fix that at the cost of a few users suffering from a loss of tabs. 

rest assured i'm doing everything i can to get to the root cause of these issues. unfortunately i don't have the resources to outsource any of this. but as @liamjohnston said, you're welcome to contribute if you're able.

**Comment by dragonspirit1185** [comment]
I would contribute if I could. I know nothing about coding.
When I mentioned fix one thing to only have a bigger issue I meant the old
issue was the sites acting up and having errors for to all the cookies and
we had to clear them.
Not so bad I didn't lose tabs.
I'm not sure what you mean by losing tabs during an update I just walked
out of the room and did something for like an hour or two and came back and
all the tabs were gone.
As far as I know nothing updated.
Once I saw you talk about another thread how you have fixed all the issues
I removed the extension and then added it again to get the latest update
even then I didn't lose tabs.
Idk but like I said I'd rather have issues with sites not working forcing
me to clear cookies than have tabs vanish.
I'll look into how to reinstate certain windows.
I'm pretty sure I tried what you're mentioning.

**Comment by bakugo** [comment]
This happened to me too today. Browser just froze while I was using it and wiped hundreds of suspended tabs, session list is empty. Thankfully I had a session backup in another extension but this is is still a serious issue.

**Comment by sclient** [comment]
Well, I just had to disable the extension in opera so I could get it to update to the new version, so that youtube would stop breaking. Lost all my suspended tabs, none are available in the recovery, and now youtube at least plays videos, but all the buttons are broken. So not good!

**Comment by deanoemcke** [comment]
@sclient for the youtube button issue, it sounds like it might be the same as this person: https://github.com/deanoemcke/thegreatsuspender/issues/713

i have not been able to replicate this, and can't see how this extension could break another website. but keen to investigate more. does refreshing the youtube page help? what happens if you disable the extension and then refresh again? (after unsuspending all your tabs first of course).

are you familiar with the chrome developer tools console? if so, can you tell me if there are any errors in the console?

**Comment by deanoemcke** [comment]
@nmhung1985 the 'before update' warning will only on a subsequent update (ie. not on this current update). are you asking if i will be incorporating features from the great discarder into this extension? or the other way around (incorporating new features from this upcoming release into the great discarder)?

The great discarder has diverged pretty significantly from this extension in terms of code, and most new features from this extension are not relevant to the great discarder. however, i do plan on doing some work on the other extension once this one is out the door.

**Comment by nmhung1985** [comment]
From the description, I thought TGD was superior and you were making it for testing, so I was just curious if you would integrate its advanced features into TGS in the future. Well, it looks like TGD and TGS will be 2 separate extensions for a long time.

However, the main thing is the TGS extension has not updated on my Chrome. Do you know the reason and what would you suggest to force the update? Of course I can just uninstall and re-install, but I'll wait for your answer before doing so.

**Comment by nmhung1985** [comment]
Oh well, I found the "Developer" method to force updating. However, it turned out TGS still could not be updated (while another extension was forced to update successfully). Guess I'll have to uninstall and re-install TGS manually.

**Comment by deanoemcke** [comment]
@nmhung1985 i'm in the middle of a selective rollout. if you wait a few more days the webstore version will update automatically. if you want it now, then uninstalling and reinstalling is the way to go. Just make sure you unsuspend or backup all your tabs first :)

**Comment by Boscop** [comment]
@deanoemcke So if we just wait, we will get the new version without losing ANY of our suspended tabs?

**Comment by deanoemcke** [comment]
@Boscop If you just wait, you will at some point be automatically updated to the latest version without any warning (assuming you are still on v6.30).

When this happens, any tabs you had suspended will disappear, leaving only those tabs that were unsuspended still showing in your browser.
After the update, the extension will attempt to recreate those suspended tabs that disappeared.
In most cases, this process works flawlessly. However, there always seems to be a few users for whom this process fails.

The best thing you can so to protect yourself is install another extension which keeps a running backup of the tabs you have open.
See this issue thread for recommendations: https://github.com/deanoemcke/thegreatsuspender/issues/742

That way, if you do suffer from a loss of tabs, you can easily recover them from an automatic backup.

**Comment by Boscop** [comment]
@deanoemcke Thanks. Yes, I'm still on 6.30. I installed "Tabs Backup & Restore" but it hasn't been updated in years. Which extension do you recommend, that will definitely prevent suspended tab loss when TGS updates?

**Comment by deanoemcke** [comment]
@Boscop i'm using "Tabs Backup & Restore" and it's working fine for me. It may not be maintained, but i assume it will continue to work fine unless Google updates chrome in such a way as to break existing extension API functionality.

**Comment by DedupOperator** [comment]
SessionBuddy -- easily convert SB sessions [TUTORIAL]
---
[As published on SessionBuddy Google Group](https://groups.google.com/forum/#!topic/sessionbuddy-discuss/qWCKH3d8DW8)

1) Click on the sessions you want to convert, or select a few from Session Buddy's left panel by holding CTRL key.

2) click on the cogwheel 'settings' icon and then Export.

3) Click: JSON > Select 'Sessions, Windows, Titles, URLs' > Ensure that 'Selected' is chosen > Save to File.

4) Open Notepad++ (download link). Open the Json file saved earlier.

5) Click CTRL+H for Search and Replace. > Under 'Search Mode' choose 'Regular Expression'.

6) Next to:
 'Find What': Copy this line: 
"url": "chrome-extension://[a-z]{32}/suspended.html#ttl=.*(http[s]?\:\/\/.[a-zA-Z0-9\.\/\_?=%&#\-\+!]+\"\,)

'Replace with:'
"url": "\1

7) Click 'Replace All' and Save the file.

8) Back to Chrome window with Session Buddy opened, Click on clogwheel icon ("settings") > Click 'Import',

9) Click 'Select File', choose the saved Json File, after the conversion. Click 'Save'.

10) You have a new unsuspended saved session on the left panel. Hooray!



* This was really quick and dirty, should work with all thegreatsuspender current versions. if someone want to make the regex better you are welcome :)
* For admins: if links for other software are not appropriate around here, you may feel free to remove them.

**Comment by ctsstc** [comment]
@deanoemcke I see the Chrome store says that the current version is: `7.0.109` from August 25, 2018, but I'm still stuck on `v6.30` I'm itching to upgrade, but don't want to screw things up. I've tried clicking Chrome's update button in the extensions but it doesn't seem to do anything unfortunately :\

**Comment by deanoemcke** [comment]
@ctsstc A new version is coming very soon and it will update _everybody_. The current version on the chrome webstore has some known issues so I've halted the rollout.

**Comment by dskrepps** [comment]
A few moments ago I was watching a youtube video when suddenly the "The Great Suspender has been updated" tab opened. Instantly all of my suspended tabs were gone, leaving only my unsuspended tabs. I clicked through to the Session Management page to find there were no sessions whatsoever.

- Running Windows 8, Chrome 72.0.3626.121 64-bit.
- I had not changed any settings from their defaults aside from probably selecting the dark theme.
- The console does not show any errors on the session management page.
- I did not receive any form of warning before the update, it was instantaneous.
- Session had somewhere between 50 to 100 tabs spanning multiple months. History not an option to recover from. I might be able to recover some via the Last Session file in Chrome's profile folder later.

Maybe you should just create a sister extension that monitors that the first is maintaining sessions properly....

**Comment by a5m0** [comment]
got the newest update and it seemed fine, but then chrome crashed a few minutes later and when I tried recover session is came back with like 1/10 of the tabs I had, then checking session buddy and it was overwritten to 2 weeks ago -_-

**Comment by deanoemcke** [comment]
@a5m0 what do you see within the session management section of the great suspender? it should have saved a session with all your open tabs pre-update.

sorry about session buddy. it seems that they stopped doing automatic session saves quite a while ago. there's a thread about better alternatives: https://github.com/deanoemcke/thegreatsuspender/issues/742

**Comment by a5m0** [comment]
@deanoemcke the session management page is empty for all 3 session types, thank you for the session buddy alternative suggestion

**Comment by quingster** [comment]
I also lost almost all of my suspended pages, and the session management doesn't have any sessions at all. I suppose I'll go through the last several months of history trying to restore everything I lost...

**Comment by a5m0** [comment]
I notice on my windows machine the session management page has lots of sessions listed, my linux machine is all blank and is the one that lost sessions.

**Comment by RezaRob** [comment]
The Great Suspender:  v7.1.0
 * Browser:  Chrome Version 73.0.3683.75 (Official Build) (64-bit)
 * Operating system:  Ubuntu 16.04.6 LTS

A lot of tabs just got closed! 

Session Management shows absolutely nothing!  All the tabs seem to be permanently gone.

I also need to disable automatic updates and there is apparently no way to do that.  

I would have to necessarily stop using this extension if it is responsible for the problem.

**Comment by deanoemcke** [comment]
It sounds like linux could be a common thread here? I'll do some investigation. 
And very sorry for all of you who are suffering from lost tabs :(

**Comment by dskrepps** [comment]
I'm on Windows 8 and also have no sessions. It's even a few days later and I look to see no sessions are saving.

**Comment by RezaRob** [comment]
> I'm on Windows 8 and also have no sessions. It's even a few days later and I look to see no sessions are saving.

Actually, I was just going to come and comment that it's strange:  Chrome programming should be independent of the underlying OS, no?!  Why would it behave differently on Linux?

**Comment by deanoemcke** [comment]
@RezaRob I agree. But you do find small differences between the chromium implementations on different operating systems. It's probably grasping at straws though.

@dskrepps Would you mind doing a little bit of debugging for me? That is probably the fastest way to get to the bottom of this. The wiki has instructions on how to enable the debugging log: https://github.com/deanoemcke/thegreatsuspender/wiki/How-to-enable-the-debugging-log

Once that's turned on, you just need to open and close a tab or two. Normally this would cause the current session to get saved. I suspect that is generating an error which will hopefully be logged to the console.

**Comment by deanoemcke** [comment]
Is there anyone else on this thread that is having the issue of no current and recent sessions being recorded in the session management view? If so, would anyone like to help with debugging it to see what is going on?

The wiki has instructions on how to enable the debugging log: https://github.com/deanoemcke/thegreatsuspender/wiki/How-to-enable-the-debugging-log

Once that's turned on, you just need to open and close a tab or two. Normally this would cause the current session to get saved. I suspect that is generating an error which will hopefully be logged to the console.

**Comment by dskrepps** [comment]
Debug log attached.

[klbibkeccnjlkjkiokjodocebajanakg-1553083818550.log](https://github.com/deanoemcke/thegreatsuspender/files/2988058/klbibkeccnjlkjkiokjodocebajanakg-1553083818550.log)

**Comment by RezaRob** [comment]
I don't see "About" on the left had side of my extensions settings page.  See image.
![sss](https://user-images.githubusercontent.com/4409171/54690962-bb2d9580-4adf-11e9-89f1-9830913748b4.jpg)

On a different note, it appears like Chrome has gotten much better at handling this sort of memory overload problem at startup.  It seems like some form of lazy-load is already being implemented and it's cleaner than before.  So, I'm not entirely sure I will need this extension.

**Comment by dskrepps** [comment]
@RezaRob You're looking at Chrome's extensions settings page, you want to look at The Great Suspender's own settings page:
![tgssettings](https://user-images.githubusercontent.com/868050/54699414-56863100-4b07-11e9-8906-5b0e847d2459.png)

**Comment by RezaRob** [comment]
Debug log attached:
[klbibkeccnjlkjkiokjodocebajanakg-1553179795481.log](https://github.com/deanoemcke/thegreatsuspender/files/2992580/klbibkeccnjlkjkiokjodocebajanakg-1553179795481.log)

**Comment by deanoemcke** [comment]
@RezaRob @dskrepps Thanks heaps for this. That's super useful. But still not enough to get to the bottom of it.
Can you try pasting this into your DevTools console, press enter, and then report back with everything it says?

```
let gsDb = await gsIndexedDb.getDb();
let results = await gsDb
  .query(gsIndexedDb.DB_CURRENT_SESSIONS)
  .all()
  .desc()
  .execute();
```

Also, within DevTools, can you go to the Application tab at the top, then expand `IndexedDB` on the sidebar. You should see `tgs - chrome-extension://...` and if you expand that, then `gsCurrentSessions`. Once you click on this, it should show you a list with one entry which is your current session. Do you have this entry?

Thanks for your help.

Also, @RezaRob, I'm aware of the improvements that chrome has made on startup. It actually starts up better with no suspended tabs than if you do have them (as I need to do extra work to make sure the suspended tabs get the correct favicons - due to the new lazy loading :/ ).
BUT, this doesn't solve the problem of chrome using up a lot of ram once you have opened many tabs. It's only good so long as you dont visit all of those tabs that got restored.

I agree though, it's one less use-case for this extension. Which is actually a good thing. Hopefully one day this extension will not be useful at all :D

**Comment by RezaRob** [comment]
@deanoemcke, I really appreciate what you're trying to do.  This extension has been a lifesaver for me for a long time, and I'm really grateful for that.  I owe you for that.  

Regarding excess memory after use, Chrome Task Manager let's you stop high memory consuming processes fairly easily.  Then, to restart them, pressing the backward/forward navigation buttons (NOT the reload button!) on the stopped page will do a reload, and gets you back to the same page position that you were reading before.  I have a feeling that'll be enough for me.

Moreover, a matter of principle bothers me:  even last time that I gave you the bug report, I lost several pages.  I don't want to run an extension which could easily threaten my browser/work state upon an automatic update which is completely out of my control.  This should be fully tested _before_ I have to do the debugging on my end.  One option would be a virtual "linux" machine to do the test.  In any case, like I said, a Chrome extension should be platform independent, and I still don't understand why it wouldn't be.  

Years ago, I had this discussion with the Chromium Dev folks (I think) on their forums, and I begged them to put these features into the official release because I knew problems like this would arise.  Now Chrome has gotten better with these memory/startup issues apparently, but perhaps it still needs some more memory management features (after startup.)

We already have a Session Manager extension which appears fairly simple and light weight, and seems to work well.  I'm not sure where it stores the session info, but something like this, I suppose, shouldn't be too difficult to do:  https://chrome.google.com/webstore/detail/session-manager/bbcnbpafconjjigibnhbfmmgdbbkcjfi

**Comment by pluto00987** [comment]
> ### What is the safe way to remove the extension?
> If you want to uninstall the extension, please _unsuspend all tabs_ before doing so. This is the only way to prevent those tabs from disappearing. This can be done easily by clicking the 'Unsuspend all tabs' option in the extension popup menu. Or more manually by visiting every suspended tab and manually reloading it.
> **Please note that if using the 'Unsuspend all tabs' option, you will need to do this once for _each chrome window_ you have open.**


FWIW I was able to set and use the extension's 'Unsuspend all tabs in all windows' hotkey to accomplish this. It was MUCH easier than manually going to each window and doing it. I was on version 6.30.

**Comment by TheKnarf** [comment]
I just lost all of my suspended tabs (some from months ago) and can't seem to get any of them back. The extension suddenly got disabled and Chrome said that it needed new right or something? Session management doesn't show anything and now I'm a bit lost..

**Comment by slfn-ldkf** [comment]
Just lost all of my suspended tabs, because this updated without asking. Session management is empty, they don't show as closed in my Chrome history. This is completely ridiculous and I'm removing this extension immediately, after using it for years. I'm so angry, this seems like basic usability.
![Screen Shot 2019-08-24 at 11 09 09 pm](https://user-images.githubusercontent.com/54474767/63637794-5f85ba80-c6c4-11e9-9b3f-40ee2dfe8cb0.png)

**Comment by hamakei** [comment]
Had the exact same problem. Extension trashed over a hundred tabs with no warning. Thanks for nothing.

**Comment by jvschiavo** [comment]
> Just lost all of my suspended tabs, because this updated without asking.

@slfn-ldkf 
There is no way for a Chrome extension to ask before updating. Couldn't you get them back with the session "How to recover lost tabs without The Great Suspender"?

**Comment by hamakei** [comment]
Nope. There is no previous session. It's completely blank.

**Comment by slfn-ldkf** [comment]
@jvschiavo Did you see the screenshot I posted? There was nothing there. I had Chrome open but was in conversation and not looking at it. When I looked back, I had a tab saying this had updated, my pinned and active tabs were still open, and everything else was gone. Obviously the first thing I did was check my history and follow the steps suggested. Nope. The tabs only show as closed _when they were last suspended_, which for some of these will be months back. Really not good enough.

**Comment by Lubaska** [comment]
It works nicely. Recovered 112 tabs a few minutes ago.

**Comment by snowman** [comment]
Hmmm. That's annoying, today I'm going to clean my extensions, and try it and notice this extension is not what I needed, then uninstall it and now I lost my 20 tabs.

**Comment by AmbivalentLife** [comment]
Created a Github account just to say that this was an **UTTER** lifesaver and saved dozens of tabs that I assumed were lost. Thank you so much!

**Comment by dm17** [comment]
Evil. 500 tabs lost. Somehow they were taken away from session buddy ability to auto save as it does? I'm just about ready to write my own; I've lost tabs under so many browsers and plugins. This is BABY LEVEL DEVELOPMENT. How is this still a common occurrence after so many years? It is called *USER DATA* - fools!

**Comment by dm17** [comment]
"chrome://history" - not true! If you haven't restarted your browser for weeks, then there is no way you'll get those tabs back without many hours of work - days if there are hundreds of tabs.

**Comment by dm17** [comment]
"I would recommend anyone wanting to remove the extension to first back up their tabs using another extension called "Session buddy". This tool will allow you to back up all your tabs and restore them again at a later date. Please be aware that those tabs suspended at the time the session buddy backup is performed will not have their correct urls. These links will only work as long as The Great Suspender is currently installed on your browser. If you want the real urls in your session buddy backup, then you will need to unsuspend all your tabs first."

Another bogus recommendation. When The Great Suspender is uninstalled. It deletes all your stuff, including recent stuff from Session Buddy.

**Comment by tusharvb19** [comment]
Any way to recover lost tabs? can we temporary enable the extension?
![image](https://user-images.githubusercontent.com/20925909/106940365-79478000-6747-11eb-8fee-231ec07833b0.png)

**Comment by clach04** [comment]
@tusharvb19 I just installed the previous version and whilst that is successful, the session data appears to be lost.

**Comment by HEYimHeroic** [comment]
> > I just lost all of my suspended tabs (some from months ago) and can't seem to get any of them back. The extension suddenly got disabled and Chrome said that it needed new right or something? Session management doesn't show anything and now I'm a bit lost..
> > > If you have access to system backups, you may be able to restore old 'recent sessions' from these backups. The recent sessions are stored in an IndexedDB database at Chrome/Default/IndexedDB/chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob/ and Chrome/Default/IndexedDB/chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.leveldb/
> > 
> > 
> > I found the folders and took a copy of them, but what do I do with them @deanoemcke ?
> > The blob directory seems to just be a bunch of empty files.
> > There seems to be a logfile in the leveldb folder called `019280.log`. Running `strings 019280.log` on it reveals I what I assume to be a log of all tabs ever suspended, but thats around ~4500 items, it doesnt seem to be any easy way of finding the set of tabs that I lost?
> > Running `strings 019280.log>tmpfile` and then opening it in Vim (`vim tmpfile`) and starting at the bottom I'm able to manually recover the tabs I've that only got suspended a few days ago. But the ones who've been there longer I imagine will either take me a long time or I'll have to give up.
> 
> I'm just trying to figure this out. I also found a log file. I didn't have a blob directory at all! How did you run strings?

unfortunately, it seems my log files were completely blank after the new update. 0 bytes. surely i can do the same process, but for the saved sessions as well?

**Comment by clach04** [comment]
Best solution I've seen so far that works (but tedious) is https://github.com/greatsuspender/thegreatsuspender/issues/1263#issuecomment-773538665

**Comment by dm17** [comment]
Can we change the name to "the great tab destroyer"?

**Comment by wchristian** [comment]
@clach04 this is much easier: https://github.com/gioxx/MarvellousSuspender/issues/7

**Comment by joerobb1** [comment]
"The Great Destroyer " got me too.  Gonna try this later this evening:  https://antimatter15.com/2015/12/recovering-deleted-data-from-leveldb/

May be able to grok out the suspended urls with this method.

**Comment by florinvirdol** [comment]
I was able to manual recover my lost suspended tabs after Chrome webstore removal of the The Great suspender extension by the following steps:

- in: `C:\Users\__YOUR_USER_NAME__\AppData\Local\Google\Chrome\User Data\Default\IndexedDB\chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob\1\__SOME_NUMBER__` i found a file (eg.: `26b3`)
- opened that file in VS Code
- a message appeared: _"The file is not displayed in the editor because it is either binary or uses an unsupported text encoding. **Do you want to open it anyway?**"_
- after opening, i searched for `uri=https://` followed by `"width"` and those were the lost suspended links

**Comment by bluesealjs** [comment]
I had over two hundreds of important tabs ..wtf, how can I recover it, chrome just auto removed the great malware.

**Comment by scarrrr316** [comment]
https://github.com/greatsuspender/thegreatsuspender/issues/1304#issuecomment-773534043
this method worked perfectly !!!!!!!  i export all my tabs through this method !!! try it

**Comment by GarlicAAZ** [comment]
I was trying to move my saved sessions from the "malware" version of TGS to the "marvellous suspender"  (should work for the "pre-malware" 7.1.6 version too) and this method worked for me:

- install the marvellous suspender (or 7.1.6 TGS)
- close chrome, then reopen it
- when chrome launches i have about 1 minute of time while the "malware" TGS version still works and chrome does not disable it (may be you will need to turn it on in the app settings), so open the TGS settings and EXPORT your sessions before the chrome disables
- turn off the "malware" TGS, turn on the new suspender app (marvellous suspender or 7.1.6) and import your exported sessions

**Comment by clach04** [comment]
Another alternative, backup history database and then export it. Basically combine

* https://github.com/greatsuspender/thegreatsuspender/issues/1263#issuecomment-773538665
* https://superuser.com/a/602274

    cd "%LocalAppData%\Google\Chrome\User Data\Default"
    copy History c:\tmp\chrome_history.sqlite3
    cd c:\tmp\
    sqlite3 chrome_history.sqlite3

Issue>

    .headers on
    .mode csv
    .output my-history.csv
    SELECT datetime(last_visit_time/1000000-11644473600,'unixepoch','localtime'), url FROM urls where url like 'chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html%' ORDER BY last_visit_time DESC;

**Comment by bluesealjs** [comment]
I don't want to use this malware anymore, what is an alternative extension and I can import the TGS saved sessions, any idea?

**Comment by bluesealjs** [comment]
This malware should be sued in the court!

**Comment by quetzalcoatl** [comment]
so... who actually is mantaining this project now?

**Comment by adamhigerd** [comment]
The successor project can be found at https://github.com/gioxx/MarvellousSuspender

**Comment by bluesealjs** [comment]
I have heard they have assigned a new maintainer so that the original maintainer won't take the blame, I guess they knew what they were doing.

**Comment by parautenbach** [comment]
> Another alternative, backup history database and then export it. Basically combine
> 
> * [#1263 (comment)](https://github.com/greatsuspender/thegreatsuspender/issues/1263#issuecomment-773538665)
> * https://superuser.com/a/602274
>   cd "%LocalAppData%\Google\Chrome\User Data\Default"
>   copy History c:\tmp\chrome_history.sqlite3
>   cd c:\tmp
>   sqlite3 chrome_history.sqlite3
> 
> Issue>
> 
> ```
> .headers on
> .mode csv
> .output my-history.csv
> SELECT datetime(last_visit_time/1000000-11644473600,'unixepoch','localtime'), url FROM urls where url like 'chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html%' ORDER BY last_visit_time DESC;
> ```

Following up on @clach04's helpful snippet and as I'm going through the process of recovery: You can run the following on the resulting CSV file to get a list of the URLs/URIs.

```
grep klbibkeccnjlkjkiokjodocebajanakg my-history.csv |grep -Eo "uri=(.+)" |cut -d = -f 2 >recovered_tabs.txt
```

There are numerous other ways to achieve the same thing using awk, sed, etc. – this one works for me. YMMV – depending on your platform. 

Once you have the file, you can check [this SO post](https://unix.stackexchange.com/questions/17659/opening-multiple-urls-from-a-text-file-as-different-tabs-in-firefox-chrome) for running through that file and open all those links. Just be mindful of not trying to open too many at once since you'll probably bring Chrome and your system to its knees – which would be the very reason you've been using the extension.

**Comment by tbal** [comment]
For me this way worked:

- started chrome
- in case some windows with supended tabs seem missing: Ctrl+Shift+T (Windows+Linux, dunno on Mac..) multiple times to reopen previously closed windows/tabs
- select every supsended tab and do a history back ("<-" arrow on the left of the url input field, ALT+(left arrow key) or Mouse Button 4 or 5); will load the actual suspended url
- installed "The Marvellous Suspender" from [chrome store](https://chrome.google.com/webstore/detail/the-marvellous-suspender/noogafoofpebimajpfpamcfhoaifemoa)
- use it as before..

EDIT: Does not work for all tabs in my case.. :|

**Comment by ghost** [comment]
This extension is removed from google it been found to contain found malware by the abuse team. google page for the extension shows a 404 error.

**Comment by johanlunds** [comment]
> I was able to manual recover my lost suspended tabs after Chrome webstore removal of the The Great suspender extension by the following steps:
> 
> * in: `C:\Users\__YOUR_USER_NAME__\AppData\Local\Google\Chrome\User Data\Default\IndexedDB\chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob\1\__SOME_NUMBER__` i found a file (eg.: `26b3`)
> * opened that file in VS Code
> * a message appeared: _"The file is not displayed in the editor because it is either binary or uses an unsupported text encoding. **Do you want to open it anyway?**"_
> * after opening, i searched for `uri=https://` followed by `"�width"` and those were the lost suspended links

Thanks, this worked great!

I used your solution and wrote this Ruby script to export the most recently saved tab URLs to a JSON file:

**MAKE A BACKUP OF THE FILES FIRST!**

```ruby
require 'json'

file = Dir['chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob/**/*']
  .find_all { |f| File.file?(f)  } # skip folders, only check actual files
  .max_by { |f| File.mtime(f) } # get the most recently modified file

res = {}

read = File.read(file)
read.force_encoding("ASCII-8BIT")
regex = /uri=(https?\:\/\/.+?)\"\x05width/n
result = read.scan(regex)
res[file] = result.flatten

IO.write('result.json', JSON.pretty_generate(res))
```

**Comment by basilisk487** [comment]
Can confirm, method recommended by @GarlicAAZ works perfectly and doesn't require going through your entire history. 
Just a couple of things: 
- If you are installing version 7.1.6, load it from source as unpacked extension using Chrome's developer mode. This is an important step, because loading an unpacked extension creates a new unique ID that would not be banned by Google.   
- Disable your internet connection to give yourself more time.

To summarize, the sequence should be the following: 
- disable internet connection
- restart chrome
- enable TGS (whatever "bad" version you have installed), go to TGS settings, export the most recent session that has about the right number of open tabs
- uninstall "bad" TGS
- enable your internet connection back
- install either 7.1.6 from source or MarvellousSuspender 
- import session, then click on "open and suspend" or "open and load". do it only ONCE, as it may seem it's not doing anything at first. 
- enjoy all your tabs back!

**Comment by rodkey** [comment]
> Any way to recover lost tabs? can we temporary enable the extension?
> ![image](https://user-images.githubusercontent.com/20925909/106940365-79478000-6747-11eb-8fee-231ec07833b0.png)

Here's a way that worked for me to recover the lost tabs:
`strings "./Library/Application Support/Google/Chrome/Default/History" | grep klbibkeccnjlkjkiokjodocebajanakg | sed 's/.*uri=//' >~/suspendedurls`
The created file suspendedurls will be in time order, most recent last.

**Comment by blidarescub** [comment]
How to recover tabs:
1. Don't touch them (no refresh, back, etc)
2. Install [Session Buddy](https://sessionbuddy.com/) 
3. Export current session to a .txt file
![image](https://user-images.githubusercontent.com/4449661/106956889-f111ac80-673f-11eb-8609-dd6f90898998.png)
4. Open the .txt file with VSCode and replace all occurrences of the following regex with 6 space characters: `.*&uri=`
![image](https://user-images.githubusercontent.com/4449661/106957290-801ec480-6740-11eb-9286-84043b1265b0.png)
5. Save the file
6. Import it back into Session Buddy and restore the session.
7. Delete The Great Suspender

**Comment by dm17** [comment]
> How to recover tabs:
> 
> 1. Don't touch them (no refresh, back, etc)
> 2. Install [Session Buddy](https://sessionbuddy.com/)
> 3. Export current session to a .txt file
>    ![image](https://user-images.githubusercontent.com/4449661/106956889-f111ac80-673f-11eb-8609-dd6f90898998.png)
> 4. Open the .txt file with VSCode and replace all occurrences of the following regex with 6 space characters: `.*&uri=`
>    ![image](https://user-images.githubusercontent.com/4449661/106957290-801ec480-6740-11eb-9286-84043b1265b0.png)
> 5. Save the file
> 6. Import it back into Session Buddy and restore the session.
> 7. Delete The Great Suspender

However, the session buddy auto-saves do not save one from The Great Destroyer's reach. I lost my 200+ tabs because I thought session buddy had them, but it seems The Great Destroyer has the ability to wipe those out too. I did export all chrome history to a CSV file after this happened, so I hope to look back at that and work through it someday. If you don't restart your browser / computer often, then it is unlikely that you'll have your tabs within the last month or so of history, however!

**Comment by andrewprofile** [comment]
#1313 try this :)

**Comment by lweeks** [comment]
> To summarize, the sequence should be the following:
> 
> * disable internet connection
> * restart chrome
> * enable TGS (whatever "bad" version you have installed), go to TGS settings, export the most recent session that has about the right number of open tabs
> * uninstall "bad" TGS
> * enable your internet connection back
> * install either 7.1.6 from source or MarvellousSuspender
> * import session, then click on "open and suspend" or "open and load". do it only ONCE, as it may seem it's not doing anything at first.
> * enjoy all your tabs back!

What I just did. This does not involve exporting or importing or editing anything. I wound up back in the same state as earlier this morning, with all my tabs and windows.

1. Ensure you have Session Buddy installed, and save your current broken session just in case.
2. Disable Internet access (pull your cable, turn off wifi, I put a firewall rule in)
3. Kill Chrome and restart it
4. Without Internet access, The Great Suspender extension works fine
5. Close all your Chrome tabs/windows so you have a clean slate.
6. Go to TGS settings, find your most recent session with all your tabs. If you're lucky, you have one that's very recent, just before Chrome disabled the extension. "Open and load" it.
7. All your windows and tabs open again, but since you have no Internet access, they'll all fail to load content other than cache. This may take a while if you have a lot -- I had 40 windows and nearly 300 tabs.
8. Once they're all opened up, save your session again in Session Buddy
9. Disable TGS, shut down Chrome
10. Enable Internet, restart Chrome, wait for your windows and tabs to reload their content either via auto-restore, or re-loading the session from Session Buddy.
11. You're now back to where you were before the extension was disabled.
12. Save your session yet again in Session Buddy just for good measure.

At this point, you can look at other extensions to manage your tabs.

**Comment by petrk94** [comment]
I got it removed 1 day ago by google, and I had more then 24 windows and more then 200 tabs open in it, its used for work.
Now its a total mess (Could Google not just disable it and inform about that issue?). The suspended.html page contains 17 MB of indexeddb data and I could find the data in 
C:\Users\User\AppData\Local\Google\Chrome\User Data\Default\IndexedDB\chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob / leveldb

But I can not access the indexeddb data as the extension is blocked and trying to import the files which are in the folder, is not working

In the chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob folder I found the folder 1\15\ and there was a file  21510 by trying to open it in Notepad++, I get this result:

![image](https://user-images.githubusercontent.com/19719052/107090850-1ca89b80-6801-11eb-9e1b-6eb6480e6ecc.png)

The problem is, I dont know in which coding it was made, so I can not extract all data. Is there a way to get all the links back from it? 
Thanks

**Comment by uecasm** [comment]
You're all making this a lot harder than it needs to be.

https://github.com/gioxx/MarvellousSuspender/issues/7

You don't need to run the old version at all, just export from the Marvellous Suspender and edit the addresses before reimporting.

**Comment by Blaisorblade** [comment]
Confused. Why export/import data? This part of the advice seems easy?
> Delete everything before the &uri= text in the address bar and the page should reload corrently.

**Comment by uecasm** [comment]
You do need something a bit more mass-automated than that when you have ~2000 tabs suspended...  (yes, I have a problem)

**Comment by verglor** [comment]
For the [Session Buddy](https://chrome.google.com/webstore/detail/session-buddy/edacconmaakjimmfgnblocblbcdcpbko) you can directly update sqlite database stored in `%LOCALAPPDATA%\Google\Chrome\User Data\Default\databases\chrome-extension_edacconmaakjimmfgnblocblbcdcpbko_0\` e.g. in [DBeaver](https://dbeaver.io).
To convert URLs to [The Marvellous Suspender](https://chrome.google.com/webstore/detail/the-marvellous-suspender/noogafoofpebimajpfpamcfhoaifemoa) just execute the following SQL commands:
```
UPDATE PreviousSessions SET windows = replace(windows , 'klbibkeccnjlkjkiokjodocebajanakg', 'noogafoofpebimajpfpamcfhoaifemoa');
UPDATE SavedSessions SET windows = replace(windows , 'klbibkeccnjlkjkiokjodocebajanakg', 'noogafoofpebimajpfpamcfhoaifemoa');
```

**Comment by tris543** [comment]
> This is the official issue for users trying to recover from tab loss due to an update of the extension, or following removal/disabling of the extension. This is the number one pain point of the extension and here I will attempt to address why this is an issue, and what you can do if affected by it.
> 
> Overview:
> Why do my tabs disappear when the extension updates or is removed?
> What is the safe way to remove the extension?
> What is the safe way to update the extension
> What should I do if I have lost tabs?
> How to recover lost tabs with The Great Suspender
> How to recover lost tabs without The Great Suspender
> My suspended tab says "This site cannot be reached"
> 
> ### Why do my tabs disappear when the extension updates or is removed?
> The Great Suspender works by redirecting a tab to a new url in order to 'suspend' it. This means that the tab is now controlled by the extension process. When the extension updates or is disabled or uninstalled, this process is killed and all tabs that belong to it are removed from the browser.
> For this reason, I try to keep the number of updates to the extension to a minimum.
> 
> The extension _does_ come with an inbuilt tab recovery system that will automatically detect and reload lost tabs in aftermath of an update or extension crash. And in the event of an update, a session restore point is automatically created in the Session History page and can be restored manually.
> 
> ### What is the safe way to remove the extension?
> If you want to uninstall the extension, please _unsuspend all tabs_ before doing so. This is the only way to prevent those tabs from disappearing. This can be done easily by clicking the 'Unsuspend all tabs' option in the extension popup menu. Or more manually by visiting every suspended tab and manually reloading it.
> Please note that if using the 'Unsuspend all tabs' option, you will need to do this once for _each chrome window_ you have open.
> 
> If you failed to unsuspend all tabs before uninstalling and have lost tabs, please refer to the section below entitled "How to recover lost tabs without The Great Suspender".
> 
> Please note, uninstalling the extension will also permanently remove all extension data including tab history and extension options. Reinstalling the extension will not enable you to do any sort of recovery.
> 
> I would recommend anyone wanting to remove the extension to first back up their tabs using another extension called "[Session buddy](https://chrome.google.com/webstore/detail/session-buddy/edacconmaakjimmfgnblocblbcdcpbko?hl=en)". This tool will allow you to back up all your tabs and restore them again at a later date. Please be aware that those tabs suspended at the time the session buddy backup is performed will not have their correct urls. These links will only work as long as The Great Suspender is currently installed on your browser. If you want the real urls in your session buddy backup, then you will need to unsuspend all your tabs first.
> 
> ### What is the safe way to update the extension?
> Unfortunately chrome does not give the user the ability to manage their own extension updates. As soon as a new release is made available on the webstore, this update is automatically pushed to users.
> 
> I have done my best to mitigate the potential for lost tabs during an update by prompting users to export a backup of their tabs before accepting the new update.
> 
> As mentioned above, a session restore point will also automatically be created to save a record of your open tabs before the update. You can then recover any lost tabs via this restore point from the Session History screen accessible from the extension Options page.
> 
> ### What should I do if I have lost tabs?
> If you have lost tabs due to the extension being removed then refer to the section below entitled "How to recover lost tabs _without_ The Great Suspender".
> 
> If you have lost tabs due to the extension being disabled, then first re-enable the extension, and then refer to the section below entitled "How to recover lost tabs with The Great Suspender".
> 
> If you have lost tabs but the extension still seems to be installed and running, then refer to the section below entitled "How to recover lost tabs with The Great Suspender".
> 
> Before continuing, it's worth checking first that you have not simply switched chrome profiles. If you have multiple chrome profiles, then each one will have a separate record of tab history.
> 
> ### How to recover lost tabs with The Great Suspender
> The extension comes with its own tab history management UI to help users recover from lost tabs. Go to the extension options page (from 'settings' in the popup or 'options' when right-clicking on the extension). Then in the settings sidebar click on 'Session management'. This will show you your most recent tab sessions. You can click on each session to see more detail on the individual windows and tabs it contains.
> 
> To reload a session, simply click the 'reload' link. This will reload all windows and tabs in an 'unsuspended' state. If your session contains a very large number of tabs, then you might instead want to click 'resuspend' which will be much faster as it reloads the tabs in a suspended state.
> 
> If for some reason the missing tabs are not in your recent sessions, then please follow the guide below for recovering lost tabs without using The Great Suspender.
> 
> If you have access to system backups, you may be able to restore old 'recent sessions' from these backups. The recent sessions are stored in an IndexedDB database at `Chrome/Default/IndexedDB/chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob/` and `Chrome/Default/IndexedDB/chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.leveldb/`
> 
> ### How to recover lost tabs without The Great Suspender
> You can attempt to recover lost tabs by using chromes in-built history page. Navigate to chrome://history in a new tab and you will be shown a list of tabs you have visited in the past grouped by date and showing the most recent at the top. Somewhere in this list you will have a record of all the tabs you lost. However it can be a bit tricky to find them as they are mixed in with all the tabs you have visited and purposely closed as well.
> 
> For example, if you opened a tab one week ago, and it got suspended and you never revisited that tab, then in chrome history, it will be grouped with all the tabs from one week ago.
> 
> You can try searching for "klbibkeccnjlkjkiokjodocebajanakg" to find tabs that were suspended. This may help narrow down the list.
> 
> If you do find a lost tab in this list, there is a chance that when you try to reopen it, it will take you to a blank page saying "This site cannot be reached". Please refer to the section below on how to recover these tabs.
> 
> ### My suspended tab says "This site cannot be reached"
> It can happen that when you open a suspended tab link, or try to unsuspend a tab, you will see a blank page with the text "This site cannot be reached". And it will have a strange url that looks something like this: `chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html#ttl=Google&uri=https://www.google.com`.
> 
> This is most likely due to the fact that you no longer have The Great Suspender installed in your browser. The easiest way to recover these tabs is to reinstall the extension, and then reload the page.
> 
> Should this fail for any reason (which would happen if you tried to open this url in another browser like firefox, or from a device that does not support extensions such as an android phone), then as a last resort you can manually edit the url to recover the tab. Delete everything before the &uri= text in the address bar and the page should reload corrently.
> ie: in the example above you would end up with `https://www.google.com`.

Thats just ridiculous. Malware is malware. I just understand how this justifies what happened. That and some users reporting otherwise. The trust is broken, theres nothing that can be done about that and speak to the 2,000,000+ users who were affected.

**Comment by antonwnk** [comment]
This was likely mentioned in some form before, but here is a VERY SIMPLE, IN-BROWSER SOLUTION:

This is just a way to automate the manual removal of everything before the "&uri=" part in the Suspender URLs. I did this in Chrome but I assume there is no issue porting it to other browsers. 

You need an extension with the chrome tabs permission enabled. In my case, this was "Clutter Free" (a quite useful one for fellow tab hoarders).

Go to the Chrome extensions page (chrome://extensions) and make sure the Developer Mode toggle is selected. Go to the card of your "`chrome.tabs ` -enabled" extension and access its "background page".

![chrome_ext](https://user-images.githubusercontent.com/23464453/107116053-2c5fc880-6871-11eb-9b21-f21c4ed098f3.png)

In the console that pops up, you can execute Javascript and use the `chrome.tabs` API.
There, this "one-liner" can be used to set the URL of all the suspender tabs to the suspended page's URL:
(⚠️remember why you had this extension installed in the first place; the following will make _all_ your ex-supender tabs active)

```javascript
chrome.tabs.query({url: "chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html*"},
                   tabs => tabs.forEach(t => {
                       actual_url = t.url.split("&uri=").slice(1).join("") 
                       chrome.tabs.update(t.id, {url: actual_url})
		   }))
```

I guess this can also be used to change the tab URL directly to the scheme of the Marvelous suspender, but I haven't installed that yet.

**Comment by bwims** [comment]
I know it's a pain when Google removes an extension on your behalf, but at least that way all Chrome users find out and are protected.

How are we Chromium users supposed to know when something like this happens?  In my case, it was because someone in my home invoked Chrome to check out a possible bug in Brave.  Otherwise we would be blissfully ignorant.

 Does Google have a mailing list we could subscribe to in order to be notified?  If not, why not?   To discourage the use of such browsers?

**Comment by AlexWayfer** [comment]
> I know it's a pain when Google removes an extension on your behalf, but at least that way all Chrome users find out and are protected.
> 
> How are we Chromium users supposed to know when something like this happens? In my case, it was because someone in my home invoked Chrome to check out a possible bug in Brave. Otherwise we would be blissfully ignorant.
> 
> Does Google have a mailing list we could subscribe to in order to be notified? If not, why not? To discourage the use of such browsers?

I have Chromium on Linux and Chrome on Windows, there is the same behavior, don't worry. I restored my tabs twice, for both of them.

**Comment by bwims** [comment]
Sorry, I should have said:  Chromium-BASED browsers.  Not all of them have
the same behaviour.   Brave  and  Sware Iron both failed to uninstall the
extension.

My question is:  How can we get alerts from Google that they have blocked
an extension?  It seems like it should be obvious, even for Chrome users.

**Comment by AlexWayfer** [comment]
> Sorry, I should have said: Chromium-BASED browsers. Not all of them have the same behaviour. Brave and Sware Iron both failed to uninstall the extension. My question is: How can we get alerts from Google that they have blocked an extension? It seems like it should be obvious, even for Chrome users.

I don't know, it's required to look at browser's extensions page with deleted ones for starting investigation and answer searching, but it's still off-topic.

**Comment by harryfear** [comment]
> 
> ![chrome_ext](https://user-images.githubusercontent.com/23464453/107116053-2c5fc880-6871-11eb-9b21-f21c4ed098f3.png)
> 
> In the console that pops up, you can execute Javascript and use the `chrome.tabs` API.
> There, this "one-liner" can be used to set the URL of all the suspender tabs to the suspended page's URL:
> (⚠️remember why you had this extension installed in the first place; the following will make _all_ your ex-supender tabs active)
> 
> ```js
> chrome.tabs.query({url: "chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html*"},
>                    tabs => tabs.forEach(t => {
>                        actual_url = t.url.split("&uri=").slice(1).join("") 
>                        chrome.tabs.update(t.id, {url: actual_url})
> 		   }))
> ```
> 

Thanks, @antonwnk, this saved tons of faffing around.

**Comment by JesusAtheist** [comment]
https://github.com/gioxx/MarvellousSuspender/issues/42

**Comment by Intechgreater** [comment]
Not sure if it's been mentioned but another option to "restore" a suspended tab is to just click the back button.

**Comment by Ingablu** [comment]
> For the [Session Buddy](https://chrome.google.com/webstore/detail/session-buddy/edacconmaakjimmfgnblocblbcdcpbko) you can directly update sqlite database stored in `%LOCALAPPDATA%\Google\Chrome\User Data\Default\databases\chrome-extension_edacconmaakjimmfgnblocblbcdcpbko_0\` e.g. in [DBeaver](https://dbeaver.io).
> To convert URLs to [The Marvellous Suspender](https://chrome.google.com/webstore/detail/the-marvellous-suspender/noogafoofpebimajpfpamcfhoaifemoa) just execute the following SQL commands:
> 
> ```
> UPDATE PreviousSessions SET windows = replace(windows , 'klbibkeccnjlkjkiokjodocebajanakg', 'noogafoofpebimajpfpamcfhoaifemoa');
> UPDATE SavedSessions SET windows = replace(windows , 'klbibkeccnjlkjkiokjodocebajanakg', 'noogafoofpebimajpfpamcfhoaifemoa');
> ```

You are really great :) I have used SQLite Expert Personal because I am dumb in IT staff but your solution is fantastic and easy for Backup Buddy users. Thank you.

**Comment by afomer** [comment]
> I got it removed 1 day ago by google, and I had more then 24 windows and more then 200 tabs open in it, its used for work.
> Now its a total mess (Could Google not just disable it and inform about that issue?). The suspended.html page contains 17 MB of indexeddb data and I could find the data in
> C:\Users\User\AppData\Local\Google\Chrome\User Data\Default\IndexedDB\chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob / leveldb
> 
> But I can not access the indexeddb data as the extension is blocked and trying to import the files which are in the folder, is not working
> 
> In the chrome-extension_klbibkeccnjlkjkiokjodocebajanakg_0.indexeddb.blob folder I found the folder 1\15\ and there was a file 21510 by trying to open it in Notepad++, I get this result:
> 
> ![image](https://user-images.githubusercontent.com/19719052/107090850-1ca89b80-6801-11eb-9e1b-6eb6480e6ecc.png)
> 
> The problem is, I dont know in which coding it was made, so I can not extract all data. Is there a way to get all the links back from it?
> Thanks

Hey, I was in your same situation, @basilisk487's answer above got me exactly all the links (You can access it here: https://github.com/greatsuspender/thegreatsuspender/issues/526#issuecomment-773571731).

One more detail. The last step, to get the actual links:
Go to The Great Suspender Settings => Sessions => (For each session, you want) Extract.
You will get all the links in a .txt file, line by line.

I hope that helps!

**Comment by ghnu** [comment]
Is there a way to recover the whitelist ?

Since Google banned 7.18, I cant invoke that any more...

**Comment by ValZapod** [comment]
Did you find a flag that will turn off this dumb protection stuff? I already blocked that antivirus integrated in Chrome... It blocked one of my extensions for downloading videos.

**Comment by afterdelight** [comment]
if you migrated to firefox and want to clean the suspended links.
open the .json file or .txt file with notepad ++

find
"url": ".*?&uri=
with regular expession enabled

replace all with
"url": "

**Comment by avipars** [comment]
https://gist.github.com/avipars/0b1c05e4a1f664967a3945bb384e55ea 

did a quick python script... def more involved than using notepad++ though... but you can modify the script to do more advanced actions
